autoload/airline/extensions/branch.vim

```diff
@@ -109,6 +110,49 @@ function! s:get_hg_branch()
   return ''
 endfunction
 
+if s:has_async
+  let s:jobs = {}
+
+  function! s:on_stdout(channel, msg) dict abort
+    let self.buf .= a:msg
+  endfunction
+
+  function! s:on_exit(channel) dict abort
+    let untracked = get(g:, 'airline#extensions#branch#notexists', g:airline_symbols.notexists)
+    if empty(self.buf)
+      let s:untracked_{self.cmd}[self.file] = ''
+    elseif (self.buf[0:1] is# '??' && self.cmd is# 'git') || (self.buf[0] is# '?' && self.cmd is# 'hg')
+      let s:untracked_{self.cmd}[self.file] = untracked
+    else
+      let s:untracked_{self.cmd}[self.file] = ''
+    endif
+  endfunction
+
+  function! s:DoAsync(cmd, file)
+    if g:airline#util#is_windows && &shell =~ 'cmd'
+      let cmd = a:cmd. shellescape(a:file)
+    else
+      let cmd = ['sh', '-c', a:cmd. shellescape(a:file)]
+    endif
+    let cmdstring = split(a:cmd)[0]
+
+    let options = {'cmd': cmdstring, 'buf': '', 'file': a:file}
+    if has_key(s:jobs, a:file)
+      if job_status(get(s:jobs, a:file)) == 'run'
+        return
+      else
+        call job_stop(get(s:jobs, a:file))
```

Comment:
According to docs, the job can be in 3 states:
`run`, `fail`, `dead`. If the job is not in `run` state, then it doesn't make sense to stop it. If you are doing this to be future-proof, in case a different state is added, then consider adding a comment.